arch/xtensa/core/gen_zsr.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 Intel corporation
+# SPDX-License-Identifier: Apache-2.0
+import sys
+import re
+
+# Scratch register allocator.  Zephyr uses multiple Xtensa SRs as
+# scratch space for various special purposes.  Unfortunately the
+# configurable architecture means that not all registers will be the
+# same on every device.  This script parses a pre-cooked ("gcc -E
+# -dM") core-isa.h file for the current architecture and assigns
+# registers to usages.
+
+NEEDED = ("ALLOCA", "CPU", "FLUSH")
+
+coreisa = sys.argv[1]
+outfile = sys.argv[2]
+
+syms = {}
+
+def get(s):
+    return syms[s] if s in syms else 0
+
+with open(coreisa) as infile:
+    for line in infile.readlines():
+        m = re.match(r"^#define\s+([^ ]+)\s*(.*)", line.rstrip())
+        if m:
+            syms[m.group(1)] = m.group(2)
+
+# Use MISC registers first if available, that's what they're for
+regs = [ f"MISC{n}" for n in range(0, int(get("XCHAL_NUM_MISC_REGS"))) ]
+
+# Next come EXCSAVE. Also record our highest non-debug interrupt level.
+maxint = 0
+for il in range(1, 1 + int(get("XCHAL_NUM_INTLEVELS"))):
+    regs.append(f"EXCSAVE{il}")
+    if il != int(get("XCHAL_DEBUGLEVEL")):
+        maxint = max(il, maxint)
+
+# Now emit our output header with the assignments we chose
+with open(outfile, "w") as f:
+    f.write("/* Generated File, see gen_zsr.py */\n")
+    f.write("#ifndef ZEPHYR_ZSR_H\n")
+    f.write("#define ZEPHYR_ZSR_H\n")
+    for i, need in enumerate(NEEDED):
+        f.write(f"# define ZSR_{need} {regs[i]}\n")
+        f.write(f"# define ZSR_{need}_STR \"{regs[i]}\"\n")
```

Comment:
I was curious if any of the NXP platforms needed to be modified in addition to the Intel platforms, so I tried building this PR for `nxp_adsp_imx8`. The generated output looks like this:
```c
/* Generated File, see gen_zsr.py */                                                                
#ifndef ZEPHYR_ZSR_H
#define ZEPHYR_ZSR_H
# define ZSR_ALLOCA MISC0
# define ZSR_ALLOCA_STR "MISC0"
# define ZSR_CPU MISC1
# define ZSR_CPU_STR "MISC1"
# define ZSR_FLUSH EXCSAVE1
# define ZSR_FLUSH_STR "EXCSAVE1"
# define ZSR_EXTRA0 EXCSAVE2
# define ZSR_EXTRA1 EXCSAVE3
# define ZSR_EXTRA2 EXCSAVE4
# define ZSR_RFI_LEVEL 3
# define ZSR_EPC EPC3
# define ZSR_EPS EPS3
#endif
```

Is the space between `#` and `define` deliberate?

Reply:
Mildly? I wanted to distinguish the macro definitions from the include guards for anyone reading the generated file, but I wasn't willing to dedicate lines in the script to emitting blanks, I guess?

Reply:
And in theory this is only touching arch-layer code and everyone should pick it up seamlessly, but yeah: it's always possible there's downstream code somewhere already stepping on these specific register assignments that will need to be aware (or potentially integrated).